libsawtooth/src/journal/state_verifier.rs

```diff
@@ -0,0 +1,337 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::error::Error;
+use std::sync::mpsc::{channel, Receiver, Sender};
+
+use crate::{
+    journal::commit_store::{ByHeightDirection, CommitStore, CommitStoreByHeightIterator},
+    protocol::block::BlockPair,
+    state::{
+        merkle::CborMerkleState, settings_view::SettingsView, state_view_factory::StateViewFactory,
+    },
+};
+use transact::{
+    database::error::DatabaseError,
+    execution::executor::ExecutionTaskSubmitter,
+    protocol::batch::Batch,
+    protocol::receipt::TransactionResult,
+    scheduler::{BatchExecutionResult, SchedulerError, SchedulerFactory},
+    state::{StateChange, Write},
+};
+
+/// Verify the state root hash of all blocks is in state and if not,
+/// reconstruct the missing state. Assumes that there are no "holes" in
+/// state, ie starting from genesis, state is present for all blocks up to some
+/// point and then not at all.
+pub fn verify_state(
+    commit_store: &CommitStore,
+    state_view_factory: &StateViewFactory,
+    initial_state_root: &str,
+    execution_task_submitter: ExecutionTaskSubmitter,
+    merkle_state: CborMerkleState,
+    scheduler_factory: Box<dyn SchedulerFactory>,
+) -> Result<(), StateVerificationError> {
+    // check if we should do state verification
+    let (start_block, prev_state_root) =
+        search_for_present_state_root(commit_store, state_view_factory, initial_state_root)?;
+
+    let start_block = match start_block {
+        Some(block) => block,
+        None => {
+            info!("Skipping state verification: chain head's state root is present");
+            return Ok(());
+        }
+    };
+
+    let previous_state_root = prev_state_root.unwrap_or_else(|| initial_state_root.to_string());
+
+    info!(
+        "Recomputing missing state from block {}",
+        start_block.header().block_num()
+    );
+
+    let blocks = commit_store.get_block_by_height_iter(
+        Some(start_block.header().block_num()),
+        ByHeightDirection::Increasing,
+    );
+
+    process_blocks(
+        &previous_state_root,
+        blocks,
+        execution_task_submitter,
+        merkle_state,
+        state_view_factory,
+        scheduler_factory,
+    )
+}
+
+/// Search through the blockstore and return a tuple containing:
+///   - the first block with a missing state root
+///   - the state root of that blocks predecessor
+fn search_for_present_state_root(
+    commit_store: &CommitStore,
+    state_view_factory: &StateViewFactory,
+    initial_state_root: &str,
+) -> Result<(Option<BlockPair>, Option<String>), StateVerificationError> {
```

Comment:
This would make more sense as an `Option<(BlockPair, String)>`

Reply:
updated